compiler/next/lib/types/BuiltinType.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/types/BuiltinType.h"
+
+namespace chpl {
+namespace types {
+
+
+void BuiltinType::canonicalizeBitWidth() {
+  switch (kind()) {
+    case BOOL:
+    case INT:
+    case UINT:
+      if (bitwidth_ == 0) {
+        if (kind() == BOOL) {
+          // OK, 'bool' is separate type from other bool(num),
+          // represent that with bit width 0.
+        } else {
+          bitwidth_ = 64; // e.g. `int` means `int(64)`.
+        }
+      } else {
+        assert(bitwidth_ == 8 || bitwidth_ == 16 ||
+               bitwidth_ == 32 || bitwidth_ == 64);
+      }
+      break;
+
+    case IMAG:
+    case REAL:
+      if (bitwidth_ == 0) {
+        bitwidth_ = 64;
+      } else {
+        assert(bitwidth_ == 32 || bitwidth_ == 64);
+      }
+      break;
+
+    case COMPLEX:
+      if (bitwidth_ == 0) {
+        bitwidth_ = 128;
+      } else {
+        assert(bitwidth_ == 64 || bitwidth_ == 128);
+      }
+      break;
+
+    default:
+      // always use 0 bitwidth
+      bitwidth_ = 0;
+  }
+}
+
+const char* BuiltinType::c_str() const {
```

Comment:
It would be good to note (via a comment?) that this will fail if the BuiltinType has not been canonicalized

Reply:
I could, but we are canonicalizing them in the constructor, so it shouldn't be possible to call this on a non-canonicalized one (unless we called it in the constructor before canonicalizing... but that's something we'd figure out if we did it).